### ISSUE TITLE
Documentation for UTF8.fold, UTF8.iter, UTF8.map etc

### ISSUE THREAD

**Issue by my-little-repository** [ocaml-batteries-team/batteries-included#640]
The functions `UTF8.fold`, `UTF8.iter`, `UTF8.map` etc do not appear in the documentation  of the [BatUTF8](http://ocaml-batteries-team.github.io/batteries-included/hdoc2/BatUTF8.html) module. I think that these functions are useful and should not be hidden in the docs. I actually discovered them through emacs auto-completion feature.

I just wrote a program processing large amounts of text and using these functions in place of the corresponding ones from the Text module gave a x4 speedup overall. So I think that they are definitely worth exposing to the users of the Battery library.

**Comment by UnixJunkie** [comment]
A patch to complete the documentation would be welcome.

**Comment by my-little-repository** [comment]
@UnixJunkie. Thanks for your answer. Nice avatar btw.

Since the documentation is autogenerated, I am not sure what should be modified. I can only guess that the two `(**/**)` in the [BatUTF8.mli](https://github.com/ocaml-batteries-team/batteries-included/blob/master/src/batUTF8.mli) file (line 158 and last line) are responsible for blocking the generation of the signatures. For the comments themselves, I think they can directly be copied from [BatText.mli](https://github.com/ocaml-batteries-team/batteries-included/blob/master/src/batText.mli), replacing the word `rope` by `utf8 string`  whenever it occurs. I am sorry, I don't know how the documentation system works.

**Comment by gasche** [comment]
Indeed, it seems that a choice was made in the past to discourage users from accessing UTF8.t directly, but use the nicer Text abstraction on top of it. You report that this abstraction layer is too costly in practice. There are two things that could be done:

- revert that choice and document the lower-level function
- optimize the abstraction layer to lessen that need

Would you be willing to contribute efforts towards one of those goals, or ideally both? I personally won't have time to invest in this in the following months, so would not do the bulk of the development effort, but I (and it seems @UnixJunkie as well) would be happy to answer your questions along the way.

After a very quick look at the situation, it seems that not all of the made-private functions are equal. I would say that "of course" `map`, `iter` and `fold` can be made public (for this, just send a pull request that displaces the documentation blocker `(**/**)` and does provide documentation for those functions), but documenting the `_unsafe` function would require more work (we would need to understand precisely when it is actually safe to use them and explain it in the documentation).

I am also surprised by the performance difference you give: my understanding (but I don't know much about the current batText implementation) was that ropes should be fast for large strings and only have a bookkeeping overhead on small strings, but you seem to be reporting otherwise. There may be some optimization work to do to avoid this overhead, but it may also be the case that your code uses performance anti-patterns of some sorts. Could you post an example (self-contained if possible) of realistic program performing the operations you are interested in, so that we could measure and experiment with potential reformulations or library optimizations? You are, of course, more than welcome to help on those efforts if you have some time to contribute.

**Comment by my-little-repository** [comment]
I will try to submit a pull request soon concerning the documentation. 

The documents I am parsing are wordlists, so the strings are short, and ropes probably don't really help.  All string operations are done through fold/map/iter and performance is as expected with BatUTF8 strings, this has been tested against some other programs in C. I'll try to provide a simple example.

**Comment by gasche** [comment]
It would be very nice if you could write a couple unit tests for the functions that are newly exposed. (Batteries has a strict policy of requiring unit tests for new functions contributed to the library (while old code is unequally covered); this particular case is a grey area as the implementation already exists and is already available.) It's very easy to write inline unit tests in comments, see in `batArray.mlv` for example.

**Comment by UnixJunkie** [comment]
@my-little-repository  There should be a release soon, in case you want to contribute some doc, don't wait long or it will be for the next release
